File: macos/Sources/PolicyChangeService.swift
import Foundation

enum PolicyChangeScope: String, CaseIterable {
    case dns, acl, firewall

    var label: String {
        switch self {
        case .dns: return "DNS"
        case .acl: return "ACL"
        case .firewall: return "防火墙"
        }
    }
}

enum PolicyChangeAction: String, CaseIterable {
    case add, update, delete, unchanged, invalid

    var label: String {
        switch self {
        case .add: return "新增"
        case .update: return "更新"
        case .delete: return "删除"
        case .unchanged: return "不变"
        case .invalid: return "无效"
        }
    }

    var actionable: Bool { self == .add || self == .update || self == .delete }
}

struct PolicyChangeItem: Identifiable {
    let id = UUID()
    var scope: PolicyChangeScope
    var action: PolicyChangeAction
    var name: String
    var details: String
    var currentID: String?
    var desiredSourceID: String?
    var actualID: String?
    var desiredIndex: Int = 0
    var desiredDNS: DNSRecord?
    var desiredPolicyJSON: String?
    var isSelected = false
    var status = "待执行"
}

struct PolicyChangePlan {
    var bundle: PolicyBundle
    var sourceURL: URL
    var synchronizeDeletes: Bool
    var items: [PolicyChangeItem]

    func count(_ action: PolicyChangeAction) -> Int { items.filter { $0.action == action }.count }
    var selectedCount: Int { items.filter(\.isSelected).count }
    var selectedDeleteCount: Int { items.filter { $0.isSelected && $0.action == .delete }.count }
}

enum PolicyBundleCodec {
    static func load(_ url: URL) throws -> PolicyBundle {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let value = try container.decode(String.self)
            let fractional = ISO8601DateFormatter()
            fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = fractional.date(from: value) { return date }
            let standard = ISO8601DateFormatter()
            if let date = standard.date(from: value) { return date }
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "无法解析日期：\(value)")
        }
        do {
            return try decoder.decode(PolicyBundle.self, from: Data(contentsOf: url))
        } catch {
            throw UniFiError.api("无法读取策略基线 JSON：\(error.localizedDescription)")
        }
    }

    static func save(_ bundle: PolicyBundle, to url: URL) throws {
        try BackupService.encoder.encode(bundle).write(to: url, options: .atomic)
    }
}

enum PolicyChangeService {
    static func buildPlan(
        bundle: PolicyBundle,
        sourceURL: URL,
        currentDNS: [DNSRecord],
        currentACL: [PolicyRule],
        currentFirewall: [PolicyRule],
        synchronizeDeletes: Bool
    ) -> PolicyChangePlan {
        var items: [PolicyChangeItem] = []
        if bundle.hasDNSSection { buildDNSItems(&items, desiredRecords: bundle.dnsRecords, currentRecords: currentDNS, synchronizeDeletes: synchronizeDeletes) }
        if bundle.hasACLSection { buildPolicyItems(&items, kind: .acl, desiredValues: bundle.aclRules, currentRules: currentACL, synchronizeDeletes: synchronizeDeletes) }
        if bundle.hasFirewallSection { buildPolicyItems(&items, kind: .firewall, desiredValues: bundle.firewallPolicies, currentRules: currentFirewall, synchronizeDeletes: synchronizeDeletes) }
        for index in items.indices where items[index].action == .add || items[index].action == .update { items[index].isSelected = true }
        items.sort {
            let left = (scopeOrder($0.scope), actionOrder($0.action), $0.desiredIndex, $0.name.lowercased())
            let right = (scopeOrder($1.scope), actionOrder($1.action), $1.desiredIndex, $1.name.lowercased())
            return left < right
        }
        return PolicyChangePlan(bundle: bundle, sourceURL: sourceURL, synchronizeDeletes: synchronizeDeletes, items: items)
    }

    static func editablePolicyJSON(kind: PolicyKind, value: JSONValue) throws -> String {
        guard var object = value.foundationValue as? [String: Any] else { throw UniFiError.api("策略项不是 JSON 对象。") }
        object.removeValue(forKey: "id")
        object.removeValue(forKey: "index")
        object.removeValue(forKey: "metadata")
        let raw = String(decoding: try JSONSerialization.data(withJSONObject: object), as: UTF8.self)
        let normalized = try UniFiPayloadValidator.policyPayload(kind, json: raw)
        return String(decoding: try JSONSerialization.data(withJSONObject: normalized, options: [.prettyPrinted, .sortedKeys]), as: UTF8.self)
    }

    static func policyID(_ value: JSONValue) -> String? {
        guard case .object(let object) = value, case .string(let id)? = object["id"] else { return nil }
        return id
    }

    static func policyName(_ value: JSONValue) -> String? {
        guard case .object(let object) = value, case .string(let name)? = object["name"] else { return nil }
        return name
    }

    private static func buildDNSItems(
        _ items: inout [PolicyChangeItem], desiredRecords: [DNSRecord], currentRecords: [DNSRecord], synchronizeDeletes: Bool
    ) {
        let currentByID = Dictionary(uniqueKeysWithValues: currentRecords.compactMap { record in record.id.map { ($0.lowercased(), record) } })
        let grouped = Dictionary(grouping: currentRecords, by: DNSImportService.identity)
        var desiredKeys = Set<String>()
        var matchedCurrent = Set<String>()
        var hasInvalid = false
        for source in desiredRecords {
            do {
                let desired = try UniFiPayloadValidator.normalizeDNS(source)
                let key = DNSImportService.identity(desired)
                guard desiredKeys.insert(key).inserted else {
                    hasInvalid = true
                    items.append(invalid(.dns, name: desired.key, details: "目标文件中存在重复 DNS 规则。"))
                    continue
                }
                let byID = desired.id.flatMap { currentByID[$0.lowercased()] }.flatMap { matchedCurrent.contains($0.stableID) ? nil : $0 }
                let current = byID ?? grouped[key]?.first(where: { !matchedCurrent.contains($0.stableID) })
                guard let current else {
                    items.append(PolicyChangeItem(scope: .dns, action: .add, name: "\(desired.typeLabel) · \(desired.key)", details: DNSImportService.describe(desired), desiredDNS: desired))
                    continue
                }
                matchedCurrent.insert(current.stableID)
                let changed = !dnsContentEquals(current, desired)
                items.append(PolicyChangeItem(
                    scope: .dns, action: changed ? .update : .unchanged,
                    name: "\(desired.typeLabel) · \(desired.key)",
                    details: changed ? "\(DNSImportService.describe(current))  →  \(DNSImportService.describe(desired))" : DNSImportService.describe(desired),
                    currentID: current.id, actualID: current.id, desiredDNS: desired,
                    status: changed ? "待执行" : "无需变更"
                ))
            } catch {
                hasInvalid = true
                items.append(invalid(.dns, name: source.key, details: error.localizedDescription))
            }
        }
        if synchronizeDeletes && !hasInvalid {
            for current in currentRecords where !matchedCurrent.contains(current.stableID) {
                items.append(PolicyChangeItem(
                    scope: .dns, action: .delete, name: "\(current.typeLabel) · \(current.key)",
                    details: DNSImportService.describe(current), currentID: current.id, actualID: current.id
                ))
            }
        }
    }

    private static func buildPolicyItems(
        _ items: inout [PolicyChangeItem], kind: PolicyKind, desiredValues: [JSONValue], currentRules: [PolicyRule], synchronizeDeletes: Bool
    ) {
        let scope: PolicyChangeScope = kind == .acl ? .acl : .firewall
        let currentUserRules = currentRules.filter(\.canModify)
        var unmatched = Set(currentUserRules.map(\.id))
        var desiredNames = Set<String>()
        var hasInvalid = false
        let ordered = desiredValues.sorted { policyIndex($0) < policyIndex($1) }
        for value in ordered {
            if let origin = policyOrigin(value), !origin.isEmpty, origin.caseInsensitiveCompare("USER_DEFINED") != .orderedSame { continue }
            let name = policyName(value) ?? "未命名策略"
            guard desiredNames.insert(name.lowercased()).inserted else {
                hasInvalid = true
                items.append(invalid(scope, name: name, details: "目标文件中存在重名的用户策略，无法安全匹配。"))
                continue
            }
            do {
                let editable = try editablePolicyJSON(kind: kind, value: value)
                let sourceID = policyID(value)
                let current = currentUserRules.first(where: { rule in
                    guard unmatched.contains(rule.id) else { return false }
                    if let sourceID, rule.id.caseInsensitiveCompare(sourceID) == .orderedSame { return true }
                    return rule.name.caseInsensitiveCompare(name) == .orderedSame
                })
                guard let current else {
                    items.append(PolicyChangeItem(
                        scope: scope, action: .add, name: name, details: describePolicy(kind, value),
                        desiredSourceID: sourceID, desiredIndex: policyIndex(value), desiredPolicyJSON: editable
                    ))
                    continue
                }
                unmatched.remove(current.id)
                let currentObject = try UniFiPayloadValidator.policyPayload(kind, json: current.editableJSON)
                let changed = !jsonEqual(editable, currentObject)
                items.append(PolicyChangeItem(
                    scope: scope, action: changed ? .update : .unchanged, name: name,
                    details: changed ? "\(current.action) / \(current.type)  →  \(describePolicy(kind, value))" : describePolicy(kind, value),
                    currentID: current.id, desiredSourceID: sourceID, actualID: current.id,
                    desiredIndex: policyIndex(value), desiredPolicyJSON: editable,
                    status: changed ? "待执行" : "无需变更"
                ))
            } catch {
                hasInvalid = true
                items.append(invalid(scope, name: name, details: error.localizedDescription))
            }
        }
        if synchronizeDeletes && !hasInvalid {
            for current in currentUserRules where unmatched.contains(current.id) {
                items.append(PolicyChangeItem(
                    scope: scope, action: .delete, name: current.name,
                    details: "\(current.action) / \(current.type) · \(current.description)",
                    currentID: current.id, actualID: current.id, desiredIndex: current.index
                ))
            }
        }
    }

    private static func dnsContentEquals(_ left: DNSRecord, _ right: DNSRecord) -> Bool {
        left.recordType.caseInsensitiveCompare(right.recordType) == .orderedSame &&
        left.key.caseInsensitiveCompare(right.key) == .orderedSame &&
        (left.recordType == "TXT" ? left.value == right.value : left.value.caseInsensitiveCompare(right.value) == .orderedSame) &&
        left.enabled == right.enabled && (left.ttl ?? 0) == (right.ttl ?? 0) &&
        (left.priority ?? 0) == (right.priority ?? 0) && (left.weight ?? 0) == (right.weight ?? 0) &&
        (left.port ?? 0) == (right.port ?? 0) &&
        left.service.caseInsensitiveCompare(right.service) == .orderedSame &&
        left.protocolName.caseInsensitiveCompare(right.protocolName) == .orderedSame
    }

    private static func jsonEqual(_ desiredJSON: String, _ currentObject: [String: Any]) -> Bool {
        guard let desired = try? JSONSerialization.jsonObject(with: Data(desiredJSON.utf8)) as? NSDictionary else { return false }
        return desired.isEqual(to: currentObject)
    }

    private static func invalid(_ scope: PolicyChangeScope, name: String, details: String) -> PolicyChangeItem {
        PolicyChangeItem(scope: scope, action: .invalid, name: name.isEmpty ? "无效项目" : name, details: details, status: "需要修正")
    }

    private static func policyIndex(_ value: JSONValue) -> Int {
        guard case .object(let object) = value, case .number(let number)? = object["index"] else { return 0 }
        return Int(number)
    }

    private static func policyOrigin(_ value: JSONValue) -> String? {
        guard case .object(let object) = value,
              case .object(let metadata)? = object["metadata"],
              case .string(let origin)? = metadata["origin"] else { return nil }
        return origin
    }

    private static func describePolicy(_ kind: PolicyKind, _ value: JSONValue) -> String {
        guard case .object(let object) = value else { return "-" }
        let action: String
        if kind == .acl, case .string(let raw)? = object["action"] { action = raw }
        else if case .object(let actionObject)? = object["action"], case .string(let raw)? = actionObject["type"] { action = raw }
        else { action = "-" }
        let type: String
        if kind == .acl, case .string(let raw)? = object["type"] { type = raw }
        else if case .object(let scope)? = object["ipProtocolScope"], case .string(let raw)? = scope["ipVersion"] { type = raw }
        else { type = "-" }
        let description: String
        if case .string(let raw)? = object["description"] { description = raw } else { description = "" }
        return "\(action) / \(type)\(description.isEmpty ? "" : " · \(description)")"
    }

    private static func scopeOrder(_ scope: PolicyChangeScope) -> Int {
        switch scope { case .dns: return 0; case .acl: return 1; case .firewall: return 2 }
    }

    private static func actionOrder(_ action: PolicyChangeAction) -> Int {
        switch action { case .invalid: return 0; case .add: return 1; case .update: return 2; case .delete: return 3; case .unchanged: return 4 }
    }
}

File: Services/PolicyChangeService.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using UniFiDnsManager.Models;

namespace UniFiDnsManager.Services;

public static class PolicyChangeService
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNameCaseInsensitive = true
    };

    static PolicyChangeService()
    {
        JsonOptions.Converters.Add(new JsonStringEnumConverter());
    }

    public static async Task<PolicyBundle> LoadBundleAsync(string path, CancellationToken cancellationToken = default)
    {
        try
        {
            await using var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);
            using var document = await JsonDocument.ParseAsync(stream, cancellationToken: cancellationToken);
            if (document.RootElement.ValueKind != JsonValueKind.Object)
                throw new ValidationException("策略文件必须是 JSON 对象。");

            var bundle = document.RootElement.Deserialize<PolicyBundle>(JsonOptions) ?? new PolicyBundle();
            bundle.HasDnsSection = document.RootElement.TryGetProperty("dns_records", out _) || document.RootElement.TryGetProperty("records", out _);
            bundle.HasAclSection = document.RootElement.TryGetProperty("acl_rules", out _);
            bundle.HasFirewallSection = document.RootElement.TryGetProperty("firewall_policies", out _);
            if (bundle.DnsRecords.Count == 0 && document.RootElement.TryGetProperty("records", out var legacyRecords) && legacyRecords.ValueKind == JsonValueKind.Array)
                bundle.DnsRecords = legacyRecords.Deserialize<List<DnsRecord>>(JsonOptions) ?? [];
            foreach (var record in bundle.DnsRecords.Where(record => record.RecordType == "SRV")) record.PopulateSrvParts();
            bundle.AclRules = CloneElements(bundle.AclRules);
            bundle.FirewallPolicies = CloneElements(bundle.FirewallPolicies);
            return bundle;
        }
        catch (JsonException exception)
        {
            throw new ValidationException($"无法读取策略 JSON：{exception.Message}");
        }
    }

    public static async Task SaveBundleAsync(string path, PolicyBundle bundle, CancellationToken cancellationToken = default)
    {
        await File.WriteAllTextAsync(path, JsonSerializer.Serialize(bundle, JsonOptions), cancellationToken);
    }

    public static PolicyChangePlan BuildPlan(
        PolicyBundle bundle,
        string sourcePath,
        IReadOnlyList<DnsRecord> currentDns,
        IReadOnlyList<OfficialPolicyRule> currentAcl,
        IReadOnlyList<OfficialPolicyRule> currentFirewall,
        bool synchronizeDeletes)
    {
        var items = new List<PolicyChangeItem>();
        if (bundle.HasDnsSection) BuildDnsItems(items, bundle.DnsRecords, currentDns, synchronizeDeletes);
        if (bundle.HasAclSection) BuildPolicyItems(items, OfficialPolicyKind.Acl, bundle.AclRules, currentAcl, synchronizeDeletes);
        if (bundle.HasFirewallSection) BuildPolicyItems(items, OfficialPolicyKind.Firewall, bundle.FirewallPolicies, currentFirewall, synchronizeDeletes);
        foreach (var item in items) item.IsSelected = item.Action is PolicyChangeAction.Add or PolicyChangeAction.Update;
        return new PolicyChangePlan
        {
            Bundle = bundle,
            SourcePath = sourcePath,
            SynchronizeDeletes = synchronizeDeletes,
            Items = items
                .OrderBy(item => ScopeOrder(item.Scope))
                .ThenBy(item => ActionOrder(item.Action))
                .ThenBy(item => item.DesiredIndex)
                .ThenBy(item => item.Name, StringComparer.OrdinalIgnoreCase)
                .ToList()
        };
    }

    public static PolicyBundle CaptureBundle(
        IUniFiClient client,
        IReadOnlyList<DnsRecord> dns,
        IReadOnlyList<OfficialPolicyRule> acl,
        IReadOnlyList<OfficialPolicyRule> firewall,
        PolicyOrderingSnapshot? aclOrdering,
        PolicyOrderingSnapshot? firewallOrdering) => new()
    {
        Target = client.Target,
        Site = client.Site,
        SiteId = client.SiteId,
        NetworkVersion = client.ApplicationVersion,
        DnsRecords = dns.Select(record => record.Clone()).ToList(),
        AclRules = acl.Select(rule => ParseElement(rule.RawResponseJson)).ToList(),
        FirewallPolicies = firewall.Select(rule => ParseElement(rule.RawResponseJson)).ToList(),
        AclOrdering = aclOrdering,
        FirewallOrdering = firewallOrdering
    };

    public static string ToEditablePolicyJson(OfficialPolicyKind kind, JsonElement element)
    {
        var node = JsonNode.Parse(element.GetRawText())?.AsObject()
            ?? throw new ValidationException("策略项不是 JSON 对象。");
        node.Remove("id");
        node.Remove("index");
        node.Remove("metadata");
        return OfficialPolicyJson.NormalizeAndValidate(kind, node.ToJsonString());
    }

    public static string? GetPolicyId(JsonElement element) => GetString(element, "id");

    public static string? GetPolicyName(JsonElement element) => GetString(element, "name");

    private static void BuildDnsItems(
        List<PolicyChangeItem> items,
        IReadOnlyList<DnsRecord> desiredRecords,
        IReadOnlyList<DnsRecord> currentRecords,
        bool synchronizeDeletes)
    {
        var currentByKey = currentRecords
            .GroupBy(ImportService.IdentityKey, ImportService.IdentityComparer)
            .ToDictionary(group => group.Key, group => group.ToList(), ImportService.IdentityComparer);
        var currentById = currentRecords
            .Where(record => !string.IsNullOrWhiteSpace(record.Id))
            .GroupBy(record => record.Id!, StringComparer.OrdinalIgnoreCase)
            .ToDictionary(group => group.Key, group => group.First(), StringComparer.OrdinalIgnoreCase);
        var desiredKeys = new HashSet<string>(ImportService.IdentityComparer);
        var matchedCurrent = new HashSet<DnsRecord>();
        var hasInvalid = false;

        foreach (var desiredSource in desiredRecords)
        {
            try
            {
                var desired = DnsValidator.Normalize(desiredSource);
                var key = ImportService.IdentityKey(desired);
                if (!desiredKeys.Add(key))
                {
                    hasInvalid = true;
                    items.Add(Invalid(PolicyChangeScope.Dns, desired.Key, "目标文件中存在重复 DNS 规则。"));
                    continue;
                }

                DnsRecord? current = null;
                if (!string.IsNullOrWhiteSpace(desired.Id) && currentById.TryGetValue(desired.Id, out var byId) && !matchedCurrent.Contains(byId))
                    current = byId;
                else if (currentByKey.TryGetValue(key, out var byKey))
                    current = byKey.FirstOrDefault(record => !matchedCurrent.Contains(record));

                if (current is null)
                {
                    items.Add(new PolicyChangeItem
                    {
                        Scope = PolicyChangeScope.Dns,
                        Action = PolicyChangeAction.Add,
                        Name = $"{desired.TypeLabel} · {desired.Key}",
                        Details = ImportService.Describe(desired),
                        DesiredDns = desired
                    });
                }
                else if (!DnsContentEquals(current, desired))
                {
                    matchedCurrent.Add(current);
                    items.Add(new PolicyChangeItem
                    {
                        Scope = PolicyChangeScope.Dns,
                        Action = PolicyChangeAction.Update,
                        Name = $"{desired.TypeLabel} · {desired.Key}",
                        Details = $"{ImportService.Describe(current)}  →  {ImportService.Describe(desired)}",
                        CurrentId = current.Id,
                        ActualId = current.Id,
                        DesiredDns = desired
                    });
                }
                else
                {
                    matchedCurrent.Add(current);
                    items.Add(new PolicyChangeItem
                    {
                        Scope = PolicyChangeScope.Dns,
                        Action = PolicyChangeAction.Unchanged,
                        Name = $"{desired.TypeLabel} · {desired.Key}",
                        Details = ImportService.Describe(desired),
                        CurrentId = current.Id,
                        ActualId = current.Id,
                        DesiredDns = desired,
                        Status = "无需变更"
                    });
                }
            }
            catch (Exception exception) when (exception is ValidationException or FormatException)
            {
                hasInvalid = true;
                items.Add(Invalid(PolicyChangeScope.Dns, desiredSource.Key, exception.Message));
            }
        }

        if (!synchronizeDeletes || hasInvalid) return;
        foreach (var current in currentRecords.Where(record => !matchedCurrent.Contains(record)))
        {
            items.Add(new PolicyChangeItem
            {
                Scope = PolicyChangeScope.Dns,
                Action = PolicyChangeAction.Delete,
                Name = $"{current.TypeLabel} · {current.Key}",
                Details = ImportService.Describe(current),
                CurrentId = current.Id,
                ActualId = current.Id
            });
        }
    }

    private static void BuildPolicyItems(
        List<PolicyChangeItem> items,
        OfficialPolicyKind kind,
        IReadOnlyList<JsonElement> desiredElements,
        IReadOnlyList<OfficialPolicyRule> currentRules,
        bool synchronizeDeletes)
    {
        var scope = kind == OfficialPolicyKind.Acl ? PolicyChangeScope.Acl : PolicyChangeScope.Firewall;
        var currentUserRules = currentRules.Where(rule => rule.CanModify).ToList();
        var unmatchedCurrentIds = currentUserRules.Select(rule => rule.Id).ToHashSet(StringComparer.OrdinalIgnoreCase);
        var desiredNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var hasInvalid = false;

        foreach (var element in desiredElements.OrderBy(element => GetInt(element, "index")))
        {
            var origin = element.TryGetProperty("metadata", out var metadata) && metadata.ValueKind == JsonValueKind.Object
                ? GetString(metadata, "origin")
                : null;
            if (!string.IsNullOrWhiteSpace(origin) && !string.Equals(origin, "USER_DEFINED", StringComparison.OrdinalIgnoreCase))
                continue;

            var name = GetPolicyName(element) ?? "未命名策略";
            var sourceId = GetPolicyId(element);
            if (!desiredNames.Add(name))
            {
                hasInvalid = true;
                items.Add(Invalid(scope, name, "目标文件中存在重名的用户策略，无法安全匹配。"));
                continue;
            }

            try
            {
                var editable = ToEditablePolicyJson(kind, element);
                var current = currentUserRules.FirstOrDefault(rule =>
                                  !string.IsNullOrWhiteSpace(sourceId) &&
                                  unmatchedCurrentIds.Contains(rule.Id) &&
                                  string.Equals(rule.Id, sourceId, StringComparison.OrdinalIgnoreCase))
                              ?? currentUserRules.FirstOrDefault(rule =>
                                  unmatchedCurrentIds.Contains(rule.Id) &&
                                  string.Equals(rule.Name, name, StringComparison.OrdinalIgnoreCase));
                if (current is null)
                {
                    items.Add(new PolicyChangeItem
                    {
                        Scope = scope,
                        Action = PolicyChangeAction.Add,
                        Name = name,
                        Details = DescribePolicy(kind, element),
                        DesiredSourceId = sourceId,
                        DesiredIndex = GetInt(element, "index"),
                        DesiredPolicyJson = editable
                    });
                }
                else
                {
                    unmatchedCurrentIds.Remove(current.Id);
                    var changed = !PolicyJsonEquals(editable, OfficialPolicyJson.NormalizeAndValidate(kind, current.ToEditableJson()));
                    items.Add(new PolicyChangeItem
                    {
                        Scope = scope,
                        Action = changed ? PolicyChangeAction.Update : PolicyChangeAction.Unchanged,
                        Name = name,
                        Details = changed ? $"{current.Action} / {current.Type}  →  {DescribePolicy(kind, element)}" : DescribePolicy(kind, element),
                        CurrentId = current.Id,
                        ActualId = current.Id,
                        DesiredSourceId = sourceId,
                        DesiredIndex = GetInt(element, "index"),
                        DesiredPolicyJson = editable,
                        Status = changed ? "待执行" : "无需变更"
                    });
                }
            }
            catch (Exception exception) when (exception is ValidationException or JsonException or InvalidOperationException)
            {
                hasInvalid = true;
                items.Add(Invalid(scope, name, exception.Message));
            }
        }

        if (!synchronizeDeletes || hasInvalid) return;
        foreach (var current in currentUserRules.Where(rule => unmatchedCurrentIds.Contains(rule.Id)))
        {
            items.Add(new PolicyChangeItem
            {
                Scope = scope,
                Action = PolicyChangeAction.Delete,
                Name = current.Name,
                Details = $"{current.Action} / {current.Type} · {current.Description}",
                CurrentId = current.Id,
                ActualId = current.Id,
                DesiredIndex = current.Index
            });
        }
    }

    private static bool DnsContentEquals(DnsRecord left, DnsRecord right) =>
        string.Equals(left.RecordType, right.RecordType, StringComparison.OrdinalIgnoreCase) &&
        string.Equals(left.Key, right.Key, StringComparison.OrdinalIgnoreCase) &&
        string.Equals(left.Value, right.Value, left.RecordType == "TXT" ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase) &&
        left.Enabled == right.Enabled &&
        left.Ttl.GetValueOrDefault() == right.Ttl.GetValueOrDefault() &&
        left.Priority.GetValueOrDefault() == right.Priority.GetValueOrDefault() &&
        left.Weight.GetValueOrDefault() == right.Weight.GetValueOrDefault() &&
        left.Port.GetValueOrDefault() == right.Port.GetValueOrDefault() &&
        string.Equals(left.Service, right.Service, StringComparison.OrdinalIgnoreCase) &&
        string.Equals(left.Protocol, right.Protocol, StringComparison.OrdinalIgnoreCase);

    private static bool PolicyJsonEquals(string left, string right)
    {
        var leftNode = JsonNode.Parse(left);
        var rightNode = JsonNode.Parse(right);
        return JsonNode.DeepEquals(leftNode, rightNode);
    }

    private static PolicyChangeItem Invalid(PolicyChangeScope scope, string? name, string details) => new()
    {
        Scope = scope,
        Action = PolicyChangeAction.Invalid,
        Name = string.IsNullOrWhiteSpace(name) ? "无效项目" : name,
        Details = details,
        Status = "需要修正"
    };

    private static string DescribePolicy(OfficialPolicyKind kind, JsonElement element)
    {
        var action = kind == OfficialPolicyKind.Acl
            ? GetString(element, "action")
            : element.TryGetProperty("action", out var actionObject) && actionObject.ValueKind == JsonValueKind.Object ? GetString(actionObject, "type") : null;
        var type = kind == OfficialPolicyKind.Acl
            ? GetString(element, "type")
            : element.TryGetProperty("ipProtocolScope", out var protocol) && protocol.ValueKind == JsonValueKind.Object ? GetString(protocol, "ipVersion") : null;
        var description = GetString(element, "description");
        return $"{action ?? "-"} / {type ?? "-"}{(string.IsNullOrWhiteSpace(description) ? string.Empty : $" · {description}")}";
    }

    private static string? GetString(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString() : null;

    private static int GetInt(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.TryGetInt32(out var number) ? number : 0;

    private static JsonElement ParseElement(string json)
    {
        using var document = JsonDocument.Parse(json);
        return document.RootElement.Clone();
    }

    private static List<JsonElement> CloneElements(IEnumerable<JsonElement> values) => values.Select(value => value.Clone()).ToList();

    private static int ScopeOrder(PolicyChangeScope scope) => scope switch
    {
        PolicyChangeScope.Dns => 0,
        PolicyChangeScope.Acl => 1,
        PolicyChangeScope.Firewall => 2,
        _ => 9
    };

    private static int ActionOrder(PolicyChangeAction action) => action switch
    {
        PolicyChangeAction.Invalid => 0,
        PolicyChangeAction.Add => 1,
        PolicyChangeAction.Update => 2,
        PolicyChangeAction.Delete => 3,
        PolicyChangeAction.Unchanged => 4,
        _ => 9
    };
}
